fullscreen/model/remove-last-manual.html

```diff
@@ -23,10 +24,15 @@
             assert_equals(document.fullscreenElement, last);
             assert_equals(event.target, last);
             last.remove();
+            // Because /last/ was removed from the top layer, only /first/
+            // remains and it becomes the new fullscreen element synchronously.
+            assert_equals(document.fullscreenElement, first);
             document.onfullscreenchange = t.step_func_done(function(event)
             {
-                assert_equals(document.fullscreenElement, first);
-                assert_equals(event.target, document);
+                // When the asynchronous steps of "exit fullscreen" ran, only
+                // /first/ was in the top layer, and thus we exited fully.
+                assert_equals(document.fullscreenElement, null);
```

Comment:
Yep, because we [collect documents to unfullscreen](https://fullscreen.spec.whatwg.org/#collect-documents-to-unfullscreen) twice, and one time is after the original fullscreen element has been removed from top layer, removing an element will effectively exit two levels of fullscreen. This is quite related to https://github.com/whatwg/fullscreen/issues/105.

I think that either fully exiting fullscreen or only calling "collect documents to unfullscreen" once are the most reasonable choices here. WDYT?

Reply:
I guess I've forgot some of the context of this PR. IIUC, we are discussing about updating the document state (e.g. fullscreen element, selector matching, etc.) synchronously in the removing steps of element, and that isn't part of the current spec at the moment?

If that's correct, I guess doing "fully exit fullscreen" doesn't help either, because that relies on "exit fullscreen a document" as well. It is also not clear to me how having "collect documents to unfullscreen" called once helps either. That function only collects document, and you may still end up popping an additional element in the document (because the original fullscreen element has been remove synchronously). Actually, how would this all ensure that we dispatch the `fullscreen` event for the element removal...

Reply:
That wasn't the original context, but yes, now this is supposed to match https://github.com/whatwg/fullscreen/pull/102 which you've already approved :)

What I would suggest here is to merge this and https://github.com/whatwg/fullscreen/pull/102, and then to treat https://github.com/whatwg/fullscreen/issues/105 as a regression from that in need of fixing.

Reply:
You're right that "fully exit fullscreen" wouldn't fix https://github.com/whatwg/fullscreen/issues/105.

The way that "collect documents to unfullscreen" could help is if we record both what documents to exit and what the fullscreen element is. If, after actually exiting fullscreen, that's no longer the fullscreen element, then we can still fire the fullscreenchange event at the document. Some finesse required, but I have something of this rough shape in an experimental patch for Blink. I'd like to handle it separately though, this set of test changes is mostly independent.

Reply:
> What I would suggest here is to merge this and whatwg/fullscreen#102, and then to treat whatwg/fullscreen#105 as a regression from that in need of fixing.

This makes sense, but... it doesn't really match the spec after whatwg/fullscreen#102 due to whatwg/fullscreen#105. It seems many of the `fullscreenchange` listener in the tests wouldn't be triggered with whatwg/fullscreen#102.

I guess... we can live with this mismatch between test and spec as far as we can resolve whatwg/fullscreen#105 soonish?

Reply:
Yeah, I need to resolve https://github.com/whatwg/fullscreen/issues/105 before I can land the top layer changes at all in Chromium.

Let me walk through the tests to see which ones don't match the spec now but will soon, and I'll rename those to tentative. Writing tests to match the somewhat broken state doesn't seem meaningful.

Reply:
Many test which expect to receive fullscreenchange event after removing the last fullscreen element probably don't match the spec. And I agree that writing tests to match a broken spec state doesn't make much sense.